File: internal/engine/markdown.go
package engine

import (
	"fmt"
	"strings"
)

func pct(n float64, digits int) string {
	return fmt.Sprintf("%.*f%%", digits, n)
}

func ratio(n *float64) string {
	if n == nil {
		return "n/d"
	}
	return fmt.Sprintf("%.1f", *n)
}

func monthsLabel(n *float64) string {
	if n == nil {
		return "n/d"
	}
	return fmt.Sprintf("%.1f meses", *n)
}

func shareOf(part, total float64) string {
	if total <= 0 {
		return "0.0%"
	}
	return pct((part/total)*100, 1)
}

func mdRow(label string, amount, total float64) string {
	return fmt.Sprintf("| %s | %s | %s |", label, FormatUSD(amount), shareOf(amount, total))
}

func ToMarkdown(r Result, spaceName string) string {
	if spaceName == "" {
		spaceName = r.Policy.Name
	}
	i := r.Inputs
	a := r.Allocation
	total := r.TotalAllocated
	band := "sin banda, no se cubrió infra"
	if r.Band != nil {
		band = r.Band.Label
	}
	var b strings.Builder
	fmt.Fprintf(&b, "# %s %s (%s)\n\n", spaceName, i.Month, i.Stage)
	fmt.Fprintf(&b, "- Banda: %s\n", band)
	fmt.Fprintf(&b, "- Cash in: %s | Infra: %s | Sobrante: %s\n", FormatUSD(i.CashInMonth), FormatUSD(i.InfraCostMonth), FormatUSD(r.Remaining))
	fmt.Fprintf(&b, "- Cash on hand al inicio: %s\n", FormatUSD(i.CashOnHandStart))
	fmt.Fprintf(&b, "- Runway: %s\n", monthsLabel(r.RunwayMonths))
	fmt.Fprintf(&b, "- EF: %s de %s (%s)\n", FormatUSD(r.EFAfter), FormatUSD(r.EFCap), pct(r.EFProgressPct, 1))
	fmt.Fprintf(&b, "- TPS arriba de 19: %s | Uptime: %s\n", pct(i.TPSPctAbove19, 1), pct(i.UptimePctMonth, 2))
	fmt.Fprintf(&b, "- Discord: %v (%v neto) | Jugadores únicos semana: %v | Concurrentes promedio: %v\n", i.DiscordMembers, i.DiscordNetGrowthMonth, i.UniquePlayersWeek, i.ConcurrentAvg)
	fmt.Fprintf(&b, "- Discord por jugador: %s\n\n", ratio(r.Health.DiscordRatio))
	b.WriteString("| Rubro | USD | % del cash in |\n| --- | ---: | ---: |\n")
	b.WriteString(mdRow("Infra", a.Infra, total) + "\n")
	b.WriteString(mdRow("EF fill", a.EFFill, total) + "\n")
	b.WriteString(mdRow("Product", a.Product, total) + "\n")
	b.WriteString(mdRow("Growth", a.Growth, total) + "\n")
	b.WriteString(mdRow("People", a.People, total) + "\n")
	b.WriteString(mdRow("Reserva de infra", a.InfraBuffer, total) + "\n")
	b.WriteString(mdRow("Sin asignar", a.Unallocated, total) + "\n\n")
	b.WriteString("## Reglas que aplicaron\n")
	if len(r.Alerts) == 0 {
		b.WriteString("- Ninguna regla especial.\n")
	} else {
		for _, al := range r.Alerts {
			fmt.Fprintf(&b, "- [%s] %s\n", al.Rule, al.Message)
		}
	}
	if notes := strings.TrimSpace(i.Notes); notes != "" {
		b.WriteString("\n## Notas\n")
		b.WriteString(notes)
		b.WriteByte('\n')
	}
	b.WriteByte('\n')
	return b.String()
}

File: src/lib/markdown.ts
/** Copy as markdown: the summary the owner pastes into notes or a doc. */

import { monthsLabel, pct, ratio, shareOf, usd } from './format.ts'
import type { Result } from './types.ts'

const row = (label: string, amount: number, total: number): string =>
  `| ${label} | ${usd(amount)} | ${shareOf(amount, total)} |`

function buildTable(result: Result): string[] {
  const a = result.allocation
  const total = result.total_allocated
  return [
    '| Rubro | USD | % del cash in |',
    '| --- | ---: | ---: |',
    row('Infra', a.infra, total),
    row('EF fill', a.ef_fill, total),
    row('Product', a.product, total),
    row('Growth', a.growth, total),
    row('People', a.people, total),
    row('Reserva de infra', a.infra_buffer, total),
    row('Sin asignar', a.unallocated, total),
  ]
}

function buildContext(result: Result): string[] {
  const i = result.inputs
  return [
    `- Banda: ${result.band === null ? 'sin banda, no se cubrió infra' : result.band.label}`,
    `- Cash in: ${usd(i.cash_in_month)} | Infra: ${usd(i.infra_cost_month)} | Sobrante: ${usd(result.remaining)}`,
    `- Cash on hand al inicio: ${usd(i.cash_on_hand_start)}`,
    `- Runway: ${monthsLabel(result.runway_months)}`,
    `- EF: ${usd(result.ef_after)} de ${usd(result.ef_cap)} (${pct(result.ef_progress_pct)})`,
    `- TPS arriba de 19: ${pct(i.tps_pct_above_19)} | Uptime: ${pct(i.uptime_pct_month, 2)}`,
    `- Discord: ${i.discord_members} (${i.discord_net_growth_month} neto) | Jugadores únicos semana: ${i.unique_players_week} | Concurrentes promedio: ${i.concurrent_avg}`,
    `- Discord por jugador: ${ratio(result.health.discord_ratio)}`,
  ]
}

export function toMarkdown(result: Result): string {
  const i = result.inputs
  const alerts = result.alerts.map((alert) => `- [${alert.rule}] ${alert.message}`)
  const notes = i.notes.trim()
  return [
    `# Novum ${i.month} (${i.stage})`,
    '',
    ...buildContext(result),
    '',
    ...buildTable(result),
    '',
    '## Reglas que aplicaron',
    ...(alerts.length > 0 ? alerts : ['- Ninguna regla especial.']),
    ...(notes.length > 0 ? ['', '## Notas', notes] : []),
    '',
  ].join('\n')
}
